manager/orchestrator/restart/restart.go

```diff
@@ -190,6 +189,87 @@ func (r *Supervisor) Restart(ctx context.Context, tx store.Tx, cluster *api.Clus
 	return nil
 }
 
+// TaskRestartDelay calculates and returns the next delay based t's RestartPolicy
+// Returns nil on error parsing the RestartPolicy
+func (r *Supervisor) TaskRestartDelay(ctx context.Context, t *api.Task) *time.Duration {
+	var restartDelay time.Duration
+	backoff := &api.BackoffPolicy{
+		Base:   defaults.Service.Task.Restart.Backoff.Base,
+		Factor: defaults.Service.Task.Restart.Backoff.Factor,
+		Max:    defaults.Service.Task.Restart.Backoff.Max,
+	}
+	if t.Spec.Restart != nil {
+		if t.Spec.Restart.Backoff != nil {
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Base); b != nil && a >= 0 {
+				backoff.Base = t.Spec.Restart.Backoff.Base
+			}
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Factor); b != nil && a >= 0 {
+				backoff.Factor = t.Spec.Restart.Backoff.Factor
+			}
+			if a, b := gogotypes.DurationFromProto(t.Spec.Restart.Backoff.Max); b != nil && a >= 0 {
+				backoff.Max = t.Spec.Restart.Backoff.Max
+			}
+		} else if t.Spec.Restart.Delay != nil {
+			// Use fixed restart delay
+			var err error
+			restartDelay, err = gogotypes.DurationFromProto(t.Spec.Restart.Delay)
+			if err != nil {
+				log.G(ctx).WithError(err).Error("invalid restart delay; using default")
+				restartDelay, err = gogotypes.DurationFromProto(defaults.Service.Task.Restart.Delay)
+				if err != nil {
+					return nil
+				}
+			}
+			return &restartDelay
+		}
+	}
+
+	// Use jittered binary exponential backoff delay
+	var base, factor, max time.Duration
+	var err error
+	if base, err = gogotypes.DurationFromProto(backoff.Base); err != nil {
+		return nil
+	}
+	if factor, err = gogotypes.DurationFromProto(backoff.Factor); err != nil {
+		return nil
+	}
+	if max, err = gogotypes.DurationFromProto(backoff.Max); err != nil {
+		return nil
+	}
+
+	var failures uint64
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	serviceID := t.ServiceID
+	tuple := orchestrator.SlotTuple{
+		Slot:      t.Slot,
+		ServiceID: t.ServiceID,
+		NodeID:    t.NodeID,
+	}
+	if r.historyByService[serviceID] != nil &&
+		r.historyByService[serviceID][tuple] != nil {
+		restartInfo := r.historyByService[serviceID][tuple]
+		failures = restartInfo.failuresSinceSuccess
+		if failures < 0 {
+			// This should never happen
+			return nil
+		}
+	} else {
+		// History is not set on the first run of a task
+		failures = 0
+	}
+
+	backoffDuration := base + factor*time.Duration(1<<(failures))
```

Comment:
lol please don't bit shift it makes my brain hurt and i'm p sure the compiler optimizes multiply and divide by 2 anyway.

Reply:
@dperny: This isn't multiplication, it's computing a power of 2. Shifting is by far the simplest way to do it.

Reply:
oh good lord i totally missed that. i can't commutativity

Reply:
in that case is there any chance of an overflow?